### PR TITLE
Added an optional max_buffer_size parameter to zip and optional initial parameter to bind!

### PR DESCRIPTION
**Pull request by zsoerenm**
Sometimes the max number of updates on a signal is known before it is zipped. In this case the zip buffer size can be predefined. This avoids the call to push! and shift! multiple times on the buffer array. Moreover, it makes me more confident to set a maximum buffer size to be sure that there is no misalignment in the code. So one can be sure that the buffer size is not continuously increasing. For now it throws an error if the buffer size is exceeded. Do you suggest another approach?

Also I set an optional initial value on bind!

### CODE REVIEW

**Comment by tshort** [comment]
I don't like the extra code complexity here. @zsoerenm, could you do some timings to show how much this PR increases speeds for this use case?

`push!` can be reasonably efficient until the array needs to be moved. It might be possible to use `sizehint!`, too. `resize!` is also an option for your approach. Another option is to consider a different datatype ([this?](http://datastructuresjl.readthedocs.io/en/latest/circ_deque.html)).

**Comment by zsoerenm** [comment]
Sorry for the long delay.
The package DataStructures actually fit perfectly. 

Right now it checks whether to use CircularDeque or Deque depending on the value of max_buffer_size:
```
pasts = map(u -> max_buffer_size != 0 ? CircularDeque{eltype(u)}(max_buffer_size) : Deque{eltype(u)}(), signals)
```
I guess this is not compiler efficient. Is there another way? Anyway this should only be called once, when the zip is created. So it should not be to bad.
  
I do not have timings right now, but as pointed out by DataStructures there is a considerable performance advantage: [Deque](http://datastructuresjl.readthedocs.io/en/latest/deque.html) and even further improvements when [CircularDeque](http://datastructuresjl.readthedocs.io/en/latest/circ_deque.html) is used.

File: test/REQUIRE [hunk 1]
```diff
@@ -1 +1,2 @@
-FactCheck
\ No newline at end of file
+FactCheck
+DataStructures
```
Comment (tshort) [inline test/REQUIRE]: This isn't needed here. If it's in the main `REQUIRE` file it doesn't need to be here.

**Review by tshort — COMMENTED** [review]
I like this approach. Other than my one REQUIRE comment, I don't have any changes. 

I also see that we are failing on master with Linux. It'll probably need some tweaks when we get closer to Julia v0.7.

**Comment by zsoerenm** [comment]
Here you go.

> I also see that we are failing on master with Linux. It'll probably need some tweaks when we get closer to Julia v0.7.

It seems like FactCheck.jl is failing. It looks like it is not maintained anymore. Do you mind switching back to Base.Test?

**Comment by tshort** [comment]
I'm fine with switching to Base.Test. The only reason I went with FactCheck is because Reactive.jl used it, and I thought it'd help to mimic their tests.

**Pull request merged**